Add image segmentation docs (#1821)
spacetx/starfish: docs/source/creating_an_image_processing_pipeline/index.rst

Segmenting Cells
----------------

Unlike single-cell RNA sequencing, image-based transcriptomics methods do not physically separate

> cells before acquiring RNA information. Therefore in order to characterize cells, the RNA must be

Reviewer: ```suggestion
cells before acquiring RNA information. Therefore, in order to characterize cells, the RNA must be
```

assigned into single cells by partitioning the image volume. Accurate unsupervised cell-segmentation
is an `open problem <https://www.kaggle.com/c/data-science-bowl-2018>`_ for all biomedical imaging
disciplines ranging from digital pathology to neuroscience.

The challenge of segmenting cells depends on the structural complexity of the sample and quality

> of images available. For example a sparse cell mono-layer with a strong cytosol stain would be

Reviewer: ```suggestion
of images available. For example, a sparse cell mono-layer with a strong cytosol stain would be
```

trivial to segment but a dense heterogeneous population of cells in 3D tissue with only a DAPI stain
can be impossible to segment perfectly. On the experimental side, selecting good cell stains and
acquiring images with low background will make segmenting a more tractable task.

There are many approaches for segmenting cells from image-based transcriptomics assays. Below are
a few methods that are implemented or integrated with starfish to output a
:py:class:`.BinaryMaskCollection`, which represents a collection of labeled objects. If you do not
know which segmentation method to use, a safe bet is to start with thresholding and watershed. On

> the other hand, if you can afford to manually define ROI masks there is no better way to

Reviewer: is ROI common enough parlance that it's not necessary to define it?

Author: Good point, I will add an entry for ROI in the glossary and link to it.

guarantee accurate segmentation.

.. note::

>     While there is no "ground truth" for cell segmentation the closest approximation is manual

Reviewer: ```suggestion
    While there is no "ground truth" for cell segmentation, the closest approximation is manual
```

The traditional method for segmenting cells in fluorescence microscopy images is to threshold the
image into foreground pixels and background pixels and then label connected foreground as
individual cells. Common issues that affect thresholding such as background noise can be corrected
by preprocessing images before thresholding and filtering connected components after. There are
`many automated image thresholding algorithms <https://imagej.net/Thresholding>`_ but currently
starfish requires manually selecting a global threshold value in :py:class:`.ThresholdBinarize`.

When overlapping cells are labeled as one connected component, they are typically segmented by
using a `distance transformation followed by the watershed algorithm <https://www.mathworks
.com/company/newsletters/articles/the-watershed-transform-strategies-for-image-segmentation
.html>`_. Watershed is a classic image processing algorithm for separating objects in images and
can be applied to all types of images. Pairing it with a distance transform is particularly
useful for segmenting convex shapes like cells.

A segmentation pipeline that consists of thresholding, connected component analysis, and watershed

> is the simplest and fastest to implement but its accuracy is highly dependent on image quality.

Reviewer: ```suggestion
is simple and fast to implement but its accuracy is highly dependent on image quality.
```

The signal-to-noise ratio of the cell stain must be high enough for minimal errors after
thresholding and binary operations. And the nuclei or cell shapes must be convex to meet the
assumptions of the distance transform or else it will over-segment. Starfish includes the basic
functions to build a watershed segmentation pipeline and a predefined :py:class:`.Watershed`
segmentation class that uses the :term:`primary images<Primary Images>` as the cell stain.

* Tutorial: :ref:`Ways to segment by thresholding and watershed<tutorial_watershed_segmentation>`

Manually Defining Cells
^^^^^^^^^^^^^^^^^^^^^^^

The most accurate but time-consuming approach is to manually segment images using a tool such as
`ROI manager <https://imagej.net/docs/guide/146-30.html#fig:The-ROI-Manager>`_ in FIJI (ImageJ). It

> is a straightforward process that starfish supports by allowing ROI set binaries to be imported as a

Reviewer: "ROI set binaries" is confusing to me.  Do you mean ROI sets saved as binary files?

Author: Yes that's what I meant... I got the language from some ROI manager documentation I can't find now but I agree it's confusing. How about "that starfish supports by importing ROI sets stored in ZIP archives".

Reviewer: sounds good.

:py:class:`.BinaryMaskCollection`. These masks can then be integrated into the pipeline for
visualization and assigning spots to cells.

* Tutorial: :ref:`Loading ImageJ ROI set<tutorial_manual_segmentation>`

Machine-Learning Methods
^^^^^^^^^^^^^^^^^^^^^^^^

Besides the two classic cell segmentation approaches mentioned above, there are machine-learning
methods that aim to replicate the accuracy of manual cell segmentation while reducing the labor
required. Machine-learning algorithms for segmentation are continually improving but there is no
perfect solution for all image types yet. These methods require training data (e.g. stained
images with manually defined labels) to train a model to predict cell or nuclei locations in test
data. There are `exceptions that don't require training on your specific data <http://www.cellpose
.org/>`_ but generally training the model is something to consider when evaluating how much time
each segmentation approach will require.

Starfish currently has built-in functionality to support `ilastik <https://www.ilastik.org/>`_, a

> segmentation toolkit that leverages machine-learning. Ilastik has a Pixel Classification

Reviewer: Is there a link to Ilastik's pixel classification workflows that might be pertinent here?

Author: Yes added
